core/service/src/lib.rs

```diff
@@ -549,6 +569,10 @@ impl<Components> Future for Service<Components> where Components: components::Co
 	type Error = ();
 
 	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		if self.should_exit.load(Ordering::Relaxed) {
```

Comment:
Is it possible to check `self.on_exit.is_live()` here instead of having a separate `should_exit` flag? Or do you not want to because the mutex that uses is more expensive than checking an atomic?